### PR TITLE
🏗 Add new task `gulp codecov-upload` to upload coverage reports

### LINKED ISSUE #20480: Codecov is posting comments on PRs even though comments are disabled
We use codecov.io for code coverage reports. In our config file, we've disabled comments:

https://github.com/ampproject/amphtml/blob/0f4870b5d84b1c165557642e966e37e4ecff33f9/.codecov.yml#L1-L4

Yet, we're seeing comments being posted on PRs. For example, https://github.com/ampproject/amphtml/pull/20424#issuecomment-456858169

### PR DESCRIPTION
**Pull request by rsimha**
This PR does the following:

- Moves code coverage uploading from `tasks/runtime-test` into a new task `gulp codecov-upload`
- Fixes a bug where ads, tests, and third party code were being factored into the calculation of total coverage
- No longer uploads when tests fail, thereby preventing [incorrect / partial reports](https://codecov.io/gh/ampproject/amphtml/) (#20480)

Partial fix for #22507
Fixes #20480

### CODE REVIEW

**Comment by rsimha** [comment]
This worked. Logs: https://travis-ci.org/ampproject/amphtml/jobs/537560929#L1636

**Comment by rsimha** [comment]
> I don't think uploading code coverage reports should be decoupled from generating these reports. Especially if it's only something that Travis does... it's more of a util function that immediately follows the report generation, not its own gulp task. What do you think?

Today, code coverage is generated and uploaded separately from two places: unit and integration tests. The server then merges the sources to generate a single report. This poses a problem when one of the two test runs fails, resulting in a noisy coverage chart for [our repo](https://codecov.io/gh/ampproject/amphtml).

image

Decoupling uploading from generating solves this problem by making sure all sources have generated coverage data before uploading them once and for all.

> Also, should I wait for this PR to merge before starting #22507?

It'll be slightly easier if this were to be merged first, but not a blocker.

**Comment by rsimha** [comment]
> Merging [#22515](https://codecov.io/gh/ampproject/amphtml/pull/22515?src=pr&el=desc) into [master](https://codecov.io/gh/ampproject/amphtml/commit/fad19d92d7eff913e05ed9faef4ac076503a274a?src=pr&el=desc) will **decrease** coverage by `6.94%`.

This is expected, since we were erroneously including test and third party directories in the coverage numbers. The new (more accurate) baseline coupled with status reports (that can be made to require a minimum patch-level coverage) will help us drive up this number.

File: build-system/tasks/codecov-upload.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+const {green, yellow, cyan} = colors;
+
+/**
+ * Uploads a single report
+ * @param {string} file
+ * @param {string} flags
+ */
+function uploadReport(file, flags) {
+  const codecovExecutable = './node_modules/.bin/codecov';
+  const codecovCmd = `${codecovExecutable} --file=${file} --flags=${flags}`;
```
Comment (danielrozenberg) [inline build-system/tasks/codecov-upload.js]: Could/should this be replaced with `npx codecov`?

Comment (rsimha) [reply build-system/tasks/codecov-upload.js]: Tried that, didn't work.

- The [official docs](https://www.npmjs.com/package/codecov#usage) recommend invoking `./node_modules/.bin/codecov`.
- There's an open issue about `npx codecov` not being reliable: https://github.com/codecov/codecov-node/issues/112

**Pull request merged**